[2025-07] Add cartGiftCardCodesRemove mutation support

juanpprieto opened the pull request:
### WHY are these changes introduced?

Addresses https://github.com/Shopify/hydrogen-internal/issues/231

<img width="886" height="414" alt="Screenshot 2025-08-27 at 11 27 16 AM" src="https://github.com/user-attachments/assets/39a4f535-b347-4de6-901c-86d3ce0ca2d9" />
<img width="928" height="620" alt="Screenshot 2025-08-27 at 11 36 57 AM" src="https://github.com/user-attachments/assets/ad7e6f77-2c06-450e-89be-f9542368d488" />
<img width="960" height="694" alt="Screenshot 2025-08-27 at 11 37 04 AM" src="https://github.com/user-attachments/assets/70b49662-ab03-4f4b-afdc-12151d2e0dbb" />

The Storefront API 2025-07 release deprecated `cartGiftCardCodesUpdate` in favor of separate `cartGiftCardCodesAdd` and `cartGiftCardCodesRemove` mutations. This change addresses a critical UX issue where users couldn't remove individual gift cards when multiple were applied, since the API only returns the last 4 digits of gift card codes for security reasons.

### WHAT is this pull request doing?

This PR adds support for the new `cartGiftCardCodesRemove` mutation to enable removing specific gift cards by their IDs.

#### Key Changes

| Component | Change | Purpose |
|-----------|--------|---------|
| **Cart Handler** | Added `removeGiftCardCodes` method | Provides API to remove gift cards by ID |
| **CartForm** | Added `GiftCardCodesRemove` action | Enables form-based gift card removal |
| **Cart Fragments** | Added `id` field to `appliedGiftCards` | Required for removal operations |
| **Skeleton Template** | Added removal UI with individual remove buttons | Allows users to remove specific gift cards |

#### Technical Details

- ✅ New mutation handler: `cartGiftCardCodesRemoveDefault`
- ✅ Maintains existing additive behavior for adding/updating gift cards
- ✅ Uses `appliedGiftCardIds` parameter (not `giftCardCodes`)
- ✅ Includes comprehensive tests without `as any` usage
- ✅ Updates both Hydrogen core and skeleton template

### HOW to test your changes?

1. Start the skeleton dev server:
   ```bash
   cd templates/skeleton
   npm run dev
   ```

2. Add a product to the cart 

3. Apply multiple gift cards 

4. Verify:
   - Individual gift cards can be removed via their remove buttons
   - Adding new gift cards is still additive (doesn't replace existing ones)
   - Cart totals update correctly after removal
   - No console errors or React hooks violations

#### Checklist

- [x] I've read the [Contributing Guidelines](https://github.com/Shopify/hydrogen/blob/main/CONTRIBUTING.md)
- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've added a [changeset](https://github.com/Shopify/hydrogen/blob/main/CONTRIBUTING.md#changesets) if this PR contains user-facing or noteworthy changes
- [x] I've added [tests](https://github.com/Shopify/hydrogen/blob/main/CONTRIBUTING.md#testing) to cover my changes
- [x] I've added or updated the documentation

juanpprieto committed:
Update package-lock.json for @shopify/cli dependency update

juanpprieto committed:
[2025-07] Fix Monorepo RR7 CI build failures and improve reliability (#3091)

* fix: Complete React Router 7 migration for Express example

The Express example was partially migrated to React Router 7 but missing critical
configuration and dependencies, causing CI failures.

Changes:
- Add react-router.config.ts with proper preset and app directory configuration
- Fix vite.config.ts to call reactRouter() without arguments (RR7 API change)
- Add @react-router/dev as devDependency for TypeScript types
- Update entry.server.tsx to remove unnecessary @remix-run/web-fetch import
  (Node 20+ provides native Response globally)
- Fix environment variable loading using getEnv() helper
- Add .gitignore for React Router generated type files
- Update tsconfig.json to exclude generated directories

The example now builds successfully with React Router 7's framework mode.

* fix: Fix CLI build checks and TypeScript configuration

Multiple CLI-specific issues were causing CI failures after recent refactoring.

Changes:
- Fix build-check.mjs to expect .jsx extensions (tsup outputs .tsx→.jsx)
- Exclude assets/routes from TypeScript checking (template files for user projects)
- Align ast-grep to version 0.33.0 to match @shopify/cli's bundled version
- Fix isHydrogenMonorepo() detection to work in both dev and npm package contexts
- Add 'as const' assertions for ast-grep 0.33.0 API compatibility

Background:
- Template files were moved to assets/routes in May 2024 and inadvertently
  included in TypeScript scope
- ast-grep 0.34.1 conflicted with @shopify/cli's bundled 0.33.0

* fix: Fix CI workflow and monorepo build orchestration

The CI workflow and Turbo configuration had multiple issues causing intermittent
build failures, particularly the "No lockfile found" error.

Changes:
- Add SHOPIFY_HYDROGEN_FLAG_LOCKFILE_CHECK=false to typecheck step in CI
  (was only on build:all, but typecheck also triggers builds via Turbo)
- Add SHOPIFY_HYDROGEN_FLAG_LOCKFILE_CHECK to turbo.json globalEnv
- Add explicit build dependencies for all examples in turbo.json
- Remove docs-preview from workspaces to prevent React Router version conflicts
- Regenerate package-lock.json for dependency resolution

The root cause was that the typecheck step depends on build tasks (per turbo.json)
but wasn't receiving the lockfile check flag, causing skeleton builds to fail.
This has been a source of CI instability since the monorepo migration.

juanpprieto committed:
chore: regenerate cookbook recipes with updated patch hashes

Regenerated all cookbook recipes to update patch file hashes after recent
changes to the skeleton template. This ensures the cookbook validation
passes in CI by keeping the patches in sync with the current state of
the skeleton template.

The regeneration updates:
- All patch files with new content hashes
- Recipe YAML files with updated patch references
- README and prompt files with current examples

This fixes the 'Validate Recipes' CI job failure.

juanpprieto committed:
[2025-07] Fix React Context error during client-side hydration (#3101)

* Fix React Context errors in Vite dev server

This fixes "Cannot read properties of null (reading 'useContext')" errors that occur when React Context hooks are used in development mode.

Root Cause:
Vite's module federation can cause multiple React instances to be loaded when @shopify/hydrogen is served as an external module. React Context requires a single React instance to function properly - when multiple instances exist, Context providers from one instance cannot be consumed by hooks from another instance, resulting in null context values.

Solution:
1. Added resolve.dedupe for react, react-dom, and @shopify/hydrogen to ensure Vite uses a single instance of these packages
2. Modified optimizeDeps.include to conditionally optimize @shopify/hydrogen:
   - In production/regular projects: Optimize @shopify/hydrogen to prevent context errors
   - In monorepo development: Skip optimization to allow live reload of source changes

This maintains development workflow in the monorepo while fixing the production build issue.

* Fix React Context error during client-side hydration

Resolves 'Cannot read properties of null (reading useContext)' error
that occurred on first page load when using CSP with nonces.

The issue was caused by NonceProvider being present during SSR but
missing during client hydration, creating a React Context mismatch.

Changes:
- Export NonceProvider from @shopify/hydrogen for client-side usage
- Add NonceProvider wrapper to skeleton's entry.client.tsx
- Simplify Vite config to improve React Context stability

kdaviduik committed:
Update Vitest from v1.0.4 to v3.2.4 across all packages (#3126)

* Update Vitest from v1.0.4 to v3.2.4 across all packages

- Upgrade vitest to ^3.2.4 in all 5 packages that use it
- Upgrade @vitest/coverage-v8 to ^3.2.4 to match vitest version
- Add vitest as devDependency to create-hydrogen package (was missing)
- Add explicit coverage provider configuration (required in v3+)
  - Set provider: 'v8' in cli and hydrogen-react vitest configs

Breaking change: Vitest v2+ requires explicit coverage provider config

* Fix Vitest v3 Mock type breaking changes

In Vitest v3, vi.fn() returns Mock<() => unknown> instead of Mock<any[], any>,
requiring explicit type assertions or type parameters for mocks with specific
return types.

Changes:
- Add type assertion for fetchWithServerCache mock in storefront.test.ts
- Simplify Promise creation in session mocks (Promise.resolve instead of new Promise)
- Fix Mock type usage in Image.test.tsx for console.warn assertions
- Add vitest/globals to tsconfig for packages using global test functions
- Keep jest types alongside vitest types for @testing-library/jest-dom compatibility

The stricter typing in v3 catches more potential issues at compile time but
requires updating existing mocks that relied on v1's permissive typing.

* Fix formatting

juanpprieto committed:
2025-07 generated types

juanpprieto committed:
Add cartGiftCardCodesRemove mutation support

- Add new cartGiftCardCodesRemove mutation handler for removing specific gift cards by ID
- Integrate removeGiftCardCodes method into cart handler
- Add GiftCardCodesRemove action to CartForm
- Update cart fragments to include appliedGiftCards.id field
- Implement gift card removal UI in skeleton template
- Maintain existing additive behavior for adding/updating gift cards
- Add tests for new mutation handler

juanpprieto committed:
[2025-06] Fix MiniOxygen/Miniflare v4 redirects handling + Stable --customer-account-push  flag (#3123)

* Fix MiniOxygen to support React Router's redirectDocument for external redirects

The Miniflare v4 update introduced a breaking change where dispatchFetch automatically
follows redirects internally, preventing external OAuth redirects from working properly.
React Router's redirectDocument() expects the browser to handle external redirects,
but Miniflare was intercepting and following them within the worker runtime.

This fix adds {redirect: 'manual'} to all dispatchFetch calls in MiniOxygen, which
prevents Miniflare from following redirects automatically. This ensures that:
- External redirects (like OAuth flows to Shopify's Customer Account API) are passed
  to the browser as intended
- React Router's redirectDocument() works correctly for cross-origin navigation
- The X-Remix-Reload-Document header is respected for full document reloads

The implementation uses Function.prototype.call to pass both arguments to dispatchFetch
while maintaining TypeScript compatibility, as the property accessor doesn't properly
expose the second parameter in TypeScript definitions.

Also adds .tryhydrogen.dev to Vite's allowedHosts for development testing.

References:
- https://github.com/cloudflare/workers-sdk/issues/5018
- https://github.com/cloudflare/workers-sdk/pull/5191

* Stabilize customer-account-push flag

- Changed from customer-account-push__unstable to customer-account-push
- Removed hidden: true to make flag visible in help
- Fixed typo: Oauth → OAuth in description
- Flag now works via CLI argument or SHOPIFY_HYDROGEN_FLAG_CUSTOMER_ACCOUNT_PUSH env var

* Fix MiniOxygen redirect handling for OAuth flows

Apply {redirect: 'manual'} to Miniflare's dispatchFetch calls to prevent automatic redirect following. This ensures external redirects (like OAuth) are passed to the browser instead of being followed internally.

Changes:
- Add redirect: 'manual' parameter to dispatchFetch in both worker and node environments
- Fix Headers.getSetCookie() usage for proper Set-Cookie header handling
- Add comprehensive tests for redirect behavior in both environments
- Stabilize customer-account-push CLI flag (remove __unstable prefix)

This fixes OAuth/PKCE authentication flows in React Router 7 where redirectDocument() and external redirects need to be handled by the browser, not the server.

* Add changeset for OAuth redirect fixes and stabilized CLI flag

* Fix intermittent CI test failures in MiniOxygen tests

- Use dynamic port allocation (port: 0) for all test servers to prevent EADDRINUSE errors
- Properly handle OS-assigned ports when using port 0
- Add server tracking and cleanup to ensure all servers are properly closed
- Fix proxy server test to use dynamic port allocation

This resolves race conditions and port conflicts that occur when tests run in parallel in CI environments.

* Fix TypeScript errors in MiniOxygen tests

- Change MiniOxygenOptions to MiniOxygenPreviewOptions in test files
- Add proper type imports for Request, Response, and DispatchFetch
- Add explicit types for onRequest callback parameters

This resolves TypeScript compilation errors that were failing CI.

juanpprieto committed:
Fix linting issues: sort exports and remove duplicate imports

juanpprieto committed:
Fix tests for cartGiftCardCodesRemove mutation support

juanpprieto committed:
Update subscriptions recipe with account/subscriptions (#3124)

kdaviduik committed:
Delete unused variable

kdaviduik commented:
# Note
- Rebased onto `2025-07-sfapi-caapi-update` after that branch got rebased with main

kdaviduik committed:
Add missing appliedGiftCards id field

rbshop committed:
fixing up after rebase

rbshop committed:
Undoing an accidental non-change from the rebase

juanpprieto committed:
Merge branch 'main' into feat-cartGiftCardCodesRemove

Resolved conflicts:
- Deleted examples/{b2b,custom-cart-method,subscriptions}/app/lib/fragments.ts
  (examples moved to cookbook in main)
- Regenerated GraphQL schemas for hydrogen-react package

The id field for appliedGiftCards is preserved in skeleton template

juanpprieto committed:
Merge branch 'main' into feat-cartGiftCardCodesRemove

juanpprieto committed:
Update lock

juanpprieto committed:
Update multipass recipe

juanpprieto committed:
fix multipass

juanpprieto committed:
Fix multipass recipe for gift card removal feature and enhance patch conflict detection

- Fix TypeScript import errors in account_.login.multipass.tsx by using standard React Router types instead of generated route types
- Regenerate all multipass recipe patches to work with the gift card removal feature added to cart.tsx
- Add comprehensive patch conflict detection to apply.ts and validate.ts that:
  - Detects .orig and .rej files created during patching
  - Shows rejected patch content for easier debugging
  - Provides clear instructions for resolving conflicts
  - Throws errors early to prevent confusing TypeScript errors
- Update enhanced error logging in validate.ts to properly decode Buffer outputs

This ensures the multipass recipe validates successfully in CI with the new cartGiftCardCodesRemove feature.

juanpprieto committed:
fix: resolve cookbook recipe validation failures for custom-cart-method and metaobjects

- Regenerated all patches for both recipes to work with gift card removal feature
- Fixed incorrect shared patch file issue (both recipes had same hash but different content)
- Added human-readable descriptions to all recipe.yaml patch entries
- Both recipes now validate successfully in CI

The root cause was that the git card removal feature (cartGiftCardCodesRemove) changed
the cart.tsx structure, causing patch conflicts in fragments.ts for both recipes.
